src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/SessionTemplateComparator.kt

```diff
@@ -0,0 +1,206 @@
+package uk.gov.justice.digital.hmpps.visitscheduler.utils
+
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.SessionDetailsDto
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.location.PermittedSessionLocationDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.category.PrisonerCategoryType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.incentive.IncentiveLevel
+import java.util.*
+import java.util.function.BiPredicate
+import java.util.stream.Collectors
+
+@Component
+class SessionTemplateComparator(private val sessionDatesUtil: SessionDatesUtil) {
+
+  private val validMatch = object : BiPredicate<String?, String?> {
+    override fun test(permittedSessionLevel: String?, prisonerLevel: String?): Boolean {
+      permittedSessionLevel?.let {
+        return it == prisonerLevel
+      }
+      // If no prison level then match
+      return true
+    }
+  }
+
+  private val multipleLocationMatcher =
+    BiPredicate<PermittedSessionLocationDto, Set<PermittedSessionLocationDto>> { primarySessionLocation, comparedSessionLocations ->
+      comparedSessionLocations.stream().anyMatch { comparedSessionLocation ->
+        validMatch.test(primarySessionLocation.levelOneCode, comparedSessionLocation.levelOneCode)
+          .and(validMatch.test(primarySessionLocation.levelTwoCode, comparedSessionLocation.levelTwoCode))
+          .and(validMatch.test(primarySessionLocation.levelThreeCode, comparedSessionLocation.levelThreeCode))
+          .and(validMatch.test(primarySessionLocation.levelFourCode, comparedSessionLocation.levelFourCode))
+      }
+    }
+
+  private fun hasOverlappingDates(
+    newSessionDetails: SessionDetailsDto,
+    existingSessionDetails: SessionDetailsDto,
+  ): Boolean {
+    val newFromDate = newSessionDetails.sessionDateRange.validFromDate
+    val existingFromDate = existingSessionDetails.sessionDateRange.validFromDate
+    val newToDate = newSessionDetails.sessionDateRange.validToDate ?: newFromDate.plusYears(500)
+    val existingToDate = existingSessionDetails.sessionDateRange.validToDate ?: newSessionDetails.sessionDateRange.validToDate?.plusDays(1) ?: newToDate
+
+    return (
+      // new from date between existing from and to dates
+      (newFromDate >= existingFromDate && newFromDate <= existingToDate) ||
+        // new to date between existing from and to dates
+        (newToDate >= existingFromDate && newToDate <= existingToDate) ||
+        // new from date less than existing from date and new to date greater than existing to date
+        (newFromDate <= existingFromDate && newToDate >= existingToDate)
+      )
+  }
```

Comment:
Would use the **in** operator here 

```
    return (
      // new from date between existing from and to dates
      (newFromDate in existingFromDate..existingToDate) ||
        // new to date between existing from and to dates
        (newToDate in existingFromDate..existingToDate) ||
        // new from date less than existing from date and new to date greater than existing to date
        (newFromDate <= existingFromDate && newToDate >= existingToDate)
      )
```

Reply:
didn't use in as it might be be less performative as it might loop through all the date entries?

Reply:
updated as it does use Comparable to check.